File: b-go/ac/T787.go
package ac

import "fmt"

// 归并排序 https://www.acwing.com/problem/content/789/
func mergeSort(q []int, l, r int) {
    if l >= r {
        return
    }
    
    mid := (l + r) >> 1
    mergeSort(q, l, mid)
    mergeSort(q, mid + 1, r)
    
    i, j, t:= l, mid + 1, make([]int, 0)
    for i <= mid && j <= r {
        if q[i] <= q[j] {
            t = append(t, q[i])
            i++
        } else {
            t = append(t, q[j])
            j++
            // res += mid - i + 1  // 有(mid - i + 1)个数和q[j]是逆序对关系
        }
    }
    
    for i <= mid {
        t = append(t, q[i])
        i++
    }
    
    for j <= r {
        t = append(t, q[j])
        j++
    }
    
    copy(q[l: r + 1], t)
}

func T787() {
    var n int
    fmt.Scan(&n)
    q := make([]int, n)
    for i := 0; i < n; i++ {
        fmt.Scan(&q[i])
    }
    mergeSort(q, 0, n - 1)
	for _, v := range q {
	    fmt.Printf("%d ", v)
	}
}

File: a-java/ac/T787.java
package ac;

import java.util.Scanner;

/**
 * 归并排序 https://www.acwing.com/problem/content/789/
 */
class T787 {
    public static void merge_sort(int[] q, int l, int r) {
        if (l >= r) return;
        
        int mid = (r + l) >> 1;
        merge_sort(q, l, mid);
        merge_sort(q, mid + 1, r);

        int k = 0, i = l, j = mid + 1;
        int[] t = new int[r - l + 1];
        while (i <= mid && j <= r) {
            if (q[i] <= q[j]) t[k++] = q[i++];
            else t[k++] = q[j++];
        }
        
        while (i <= mid) t[k++] = q[i++];
        while (j <= r) t[k++] = q[j++];
        
        for (i = l, j = 0; i <= r; i++, j++) {
            q[i] = t[j];
        }
    }
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        int n = sc.nextInt();
        int[] q = new int[n];
        for (int i = 0; i < n; i++) {
            q[i] = sc.nextInt();
        }

        merge_sort(q, 0, n - 1);

        for (int i = 0; i < n; i++) {
            System.out.print(q[i] + " ");
        }
        sc.close();
    }
}
